resources/bundles/org.eclipse.core.resources/src/org/eclipse/core/internal/events/AutoBuildJob.java

```diff
@@ -298,6 +297,18 @@ private synchronized boolean shouldBuild() {
 		}
 	}
 
+	/**
+	 * Returns {@code true}, if the auto-build job should be interrupted. The job
+	 * should always be interrupted, unless this method is called by a job within
+	 * the {@link ResourcesPlugin#FAMILY_AUTO_BUILD} family.
+	 */
+	private synchronized boolean shouldInterrupt() {
+		Job currentJob = Job.getJobManager().currentJob();
+
+		// make sure autobuild doesn't interrupt itself
+		return currentJob == null || !currentJob.belongsTo(ResourcesPlugin.FAMILY_AUTO_BUILD);
```

Comment:
note, that currentJob == null check is basically an invite to modify resources from **any** non-job thread, and the following family check works not only for the actual autobuild job but for **any** job that decides to be in same family.

Reply:
> note, that currentJob == null check is basically an invite to modify resources from any non-job thread

I don't think that's the case. If `currentJob == null`, then the interrupt flag is set.
Which is how it's currently working with the `Job.getJobManager().currentJob() == this` check.

>  the following  family check works not only for the actual autobuild job but for any job that decides to be in same family.

That's the general idea. But going over the family is also how other frameworks like XText or the Debug UI plugin observe the state of the auto-builder, given that the BuildManager class is only an internal class.

Obviously it should only be used when necessary, but I think this is an appropriate use case.